cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatDataset.java

```diff
@@ -0,0 +1,205 @@
+    }
+  }
+
+  /**
+   * Write {@link RunRecordMeta} to heart beat dataset as value,
+   * with rowKey in format [namespace][row-key-separator][timestamp-in-seconds][program-run-Id]
+   *
+   * @param runRecordMeta row value to write
+   * @param timestampInSeconds used for creating rowKey
+   */
+  public void writeRunRecordMeta(RunRecordMeta runRecordMeta, long timestampInSeconds) {
+    byte[] rowKey = createRowKey(timestampInSeconds, runRecordMeta.getProgramRunId());
+    table.put(rowKey, COLUMN_NAME, Bytes.toBytes(GSON.toJson(runRecordMeta)));
+  }
+
+  private byte[] createRowKey(long timestampInSeconds, ProgramRunId programRunId) {
+    MDSKey.Builder builder = new MDSKey.Builder();
+    // add namespace at the beginning
+    builder.add(programRunId.getNamespace());
+    // add row key separator
+    // we add a rowkey separator between namespace and timestamp as namespace has variable length and there could also
+    // be namespace which might have timestamp as its name,
+    // having a separator helps to provide accurate result in those scenarios
+    builder.add(ROWKEY_SEPARATOR);
+    // add timestamp
+    builder.add(timestampInSeconds);
+    // add program runId fields
+    builder.add(programRunId.getNamespace());
+    builder.add(programRunId.getApplication());
+    builder.add(programRunId.getType().getCategoryName());
+    builder.add(programRunId.getProgram());
+    builder.add(programRunId.getRun());
+
+    return builder.build().getKey();
+  }
+
+  /**
+   * Add namespace and timestamp separated by rowKey separator and return it as the scan key
+   * @return scan key
+   */
+  private byte[] getScanKey(String namespace, long timestamp) {
+    MDSKey.Builder scanKeyBuilder = new MDSKey.Builder();
+    scanKeyBuilder.add(namespace);
+    scanKeyBuilder.add(ROWKEY_SEPARATOR);
+    scanKeyBuilder.add(timestamp);
+    return scanKeyBuilder.build().getKey();
+  }
+
+  /**
+   * Scan the table for the time range for each of the namespace provided
+   * and return collection of latest {@link RunRecordMeta}
+   * we maintain the latest {@link RunRecordMeta} identified by {@link ProgramRunId},
+   * Since there can be more than one RunRecordMeta for the
+   * same runId due to multiple state changes and heart beat messages.
+   * @param startTimestampInSeconds inclusive start rowKey
+   * @param endTimestampInSeconds exclusive end rowKey
+   * @param namespaces set of namespaces
+   * @return collection of {@link RunRecordMeta}
+   */
+  public Collection<RunRecordMeta> scan(long startTimestampInSeconds, long endTimestampInSeconds,
+                                        Set<String> namespaces) {
+    List<RunRecordMeta> resultRunRecordList = new ArrayList<>();
+    for (String namespace : namespaces) {
+      byte[] startRowKey = getScanKey(namespace, startTimestampInSeconds);
+      byte[] endRowKey = getScanKey(namespace, endTimestampInSeconds);
+      resultRunRecordList.addAll(performScan(startRowKey, endRowKey));
```

Comment:
this means it will iterate twice over everything, once in performScan() and another time to add to this list. Better to pass in the list and just add it in performScan()